core/src/main/java/org/apache/iceberg/catalog/BaseSessionCatalog.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.catalog;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.exceptions.NamespaceNotEmptyException;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableSet;
+
+public abstract class BaseSessionCatalog implements SessionCatalog {
+  private final Cache<String, Catalog> catalogs = Caffeine.newBuilder()
+      .expireAfterAccess(10, TimeUnit.MINUTES)
+      .build();
+
+  private String name = null;
+  private Map<String, String> properties = null;
+
+  @Override
+  public void initialize(String catalogName, Map<String, String> props) {
+    this.name = catalogName;
+    this.properties = props;
+  }
+
+  @Override
+  public String name() {
+    return name;
+  }
+
+  @Override
+  public Map<String, String> properties() {
+    return properties;
+  }
+
+  public Catalog asCatalog(SessionContext context) {
+    return catalogs.get(context.sessionId(), id -> new AsCatalog(context));
+  }
+
+  public <T> T withContext(SessionContext context, Function<Catalog, T> task) {
+    return task.apply(asCatalog(context));
+  }
+
+  public class AsCatalog implements Catalog, SupportsNamespaces {
```

Comment:
Can this inner class be static?  Caches historically have reference problems with non-static inner classes (though I don't recall the exact scenarios).

Reply:
I don't think this usage has the issues I was concerned about. Resolving.